### PR TITLE
termux-info: check if termux plugins are installed

### PR DESCRIPTION
**Pull request by 2096779623**

### CODE REVIEW

**Comment by 2096779623** [comment]
> `pm list packages` doesn't work on Android 13

```bash
~ $ pm list packages|grep com.termux.
package:com.termux.api
package:com.termux.styling
~ $ termux-info
Termux Variables:
TERMUX_API_APP__VERSION_NAME=0.50.1+2c6674f
TERMUX_APP_PACKAGE_MANAGER=apt
TERMUX_APP__AM_SOCKET_SERVER_ENABLED=true
TERMUX_APP__APK_PATH=/data/app/~~0NZTxnn8gRJzq6TrrRHsLQ==/com.termux-WFA1nFFqsCUVMO37JW8RqQ==/base.apk
TERMUX_APP__APK_RELEASE=GITHUB
TERMUX_APP__FILES_DIR=/data/user/0/com.termux/files
TERMUX_APP__IS_DEBUGGABLE_BUILD=true
TERMUX_APP__IS_INSTALLED_ON_EXTERNAL_STORAGE=false
TERMUX_APP__PACKAGE_MANAGER=apt
TERMUX_APP__PACKAGE_NAME=com.termux
TERMUX_APP__PACKAGE_VARIANT=apt-android-7
TERMUX_APP__PID=15239
TERMUX_APP__SE_FILE_CONTEXT=u:object_r:app_data_file:s0:c185,c256,c512,c768
TERMUX_APP__SE_INFO=default:targetSdkVersion=28:complete
TERMUX_APP__SE_PROCESS_CONTEXT=u:r:untrusted_app_27:s0:c185,c256,c512,c768
TERMUX_APP__TARGET_SDK=28
TERMUX_APP__UID=10185
TERMUX_APP__USER_ID=0
TERMUX_APP__VERSION_CODE=118
TERMUX_APP__VERSION_NAME=0.118.0
TERMUX_MAIN_PACKAGE_FORMAT=debian
TERMUX_VERSION=0.118.0
Packages CPU architecture:
aarch64
Subscribed repositories:
# sources.list
deb https://mirrors.ustc.edu.cn/termux/apt/termux-main stable main
# tur-repo (sources.list.d/tur.list)
deb https://tur.kcubeterm.com tur-packages tur tur-on-device tur-continuous
# x11-repo (sources.list.d/x11.list)
deb https://mirrors.ustc.edu.cn/termux/apt/termux-x11 x11 main
Updatable packages:
clang/stable 16.0.4 aarch64 [upgradable from: 16.0.3]
curl/stable 8.1.0 aarch64 [upgradable from: 8.0.1]
fribidi/stable 1.0.13 aarch64 [upgradable from: 1.0.12]
libcompiler-rt/stable 16.0.4 aarch64 [upgradable from: 16.0.3]
libcurl/stable 8.1.0 aarch64 [upgradable from: 8.0.1]
libllvm/stable 16.0.4 aarch64 [upgradable from: 16.0.3]
lld/stable 16.0.4 aarch64 [upgradable from: 16.0.3]
llvm/stable 16.0.4 aarch64 [upgradable from: 16.0.3]
termux-tools version:
1.38.1
Android version:
13
Kernel build information:
Linux localhost 4.14.299-Perf #4 SMP PREEMPT Sun Dec 4 15:42:55 CST 2022 aarch64 Android
Device manufacturer:
LGE
Device model:
LM-G820N
LD Variables:
LD_LIBRARY_PATH=
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.so
~ $
```
Work fine on my Android 13 🤔 .

**Comment by 2096779623** [comment]
> Do you have special permissions granted through ADB?
No.

**Comment by agnostic-apollo** [comment]
>Nothing is shown in logcat regarding pm failure

It's an selinux restriction that can be bypassed as mentioned in my comment above by using a sub shell and redirecting stdin to `/dev/null` and stderr to stout.

**Comment by agnostic-apollo** [comment]
Also do not hardcode `com.termux` in grep.

**Comment by 2096779623** [comment]
> Then perhaps it depends on manufacturer (ROM configuration) and can't be considered as stable way to retrieve list of the applications.

Please check if it works fine.

File: scripts/termux-info.in [hunk 1]
```diff
@@ -116,7 +116,12 @@ Device model:
...
 LD Variables:
 LD_LIBRARY_PATH=$LD_LIBRARY_PATH
-LD_PRELOAD=$LD_PRELOAD"
+LD_PRELOAD=$LD_PRELOAD
+$(
+TERMUX_PLUGINS=$(pm list packages | grep -v '^cmd:' | grep $termux_app_package. | cut -d ":" -f 2)
+if [ -n "$TERMUX_PLUGINS" ]; then
+        echo -e "Installed termux plugins: \n$TERMUX_PLUGINS"
+fi)"
 echo "$output"
```
[the remark this answers is not in the document]
Comment (2096779623) [reply scripts/termux-info.in:125]: > Now is better, but I have an extra empty line in output after `LD_PRELOAD`. I recommend moving the Termux plugins output after `echo "$output"`.

```suggestion
LD_PRELOAD=$LD_PRELOAD
$(
TERMUX_PLUGINS=$(pm list packages | grep -v '^cmd:' | grep $termux_app_package. | cut -d ":" -f 2)
if [ -n "$TERMUX_PLUGINS" ]; then
        echo -e "Installed termux plugins: \n$TERMUX_PLUGINS"
fi)"
echo "$output" | tr -s '\n'
```
There are still blank lines after modifying `echo "$output"`.

[the remark this answers is not in the document]
Comment (2096779623) [reply scripts/termux-info.in]: > This needs to be tested:

```bash
LD_PRELOAD=/data/data/com.termux/files/usr/lib/libtermux-exec.soInstalled termux plugins:
com.termux.api
com.termux.styling
~/termux-tools $
```
Missing a newline(`\n`), I'll add it later.

**Comment by agnostic-apollo** [comment]
The regex will match other package names as well. Wait for full response.

**Comment by 2096779623** [comment]
> The regex will match other package names as well. Wait for full response.

For example?

File: scripts/termux-info.in [hunk 2]
```diff
@@ -117,6 +117,10 @@ $(getprop ro.product.model)
 LD Variables:
 LD_LIBRARY_PATH=$LD_LIBRARY_PATH
 LD_PRELOAD=$LD_PRELOAD"
+TERMUX_PLUGINS=$(pm list packages | grep -v '^cmd:' | grep "@TERMUX_APP_PACKAGE@." | cut -d ":" -f 2)
```
[the remark this answers is not in the document]
Comment (agnostic-apollo) [reply scripts/termux-info.in:120]: @2096779623

**Comment by agnostic-apollo** [comment]
Hopefully, should have gotten a notification now for the comment above.

**Comment by agnostic-apollo** [comment]
Okay, there seems to be something wrong, my review comments are not being shown while sylirre's are in incognito mode. Maybe that's why you didn't see them before.

**Comment by agnostic-apollo** [comment]
Sorry, forgot to submit review.

**Review by agnostic-apollo — APPROVED** [review]

**Comment by agnostic-apollo** [comment]
p.s The `--show-versioncode` option was added  in Android `8` but wasn't shown in help output till android `10`, can be added if needed by checking `ro.build.sdk.version` against `>= 26`.

https://cs.android.com/android/_/android/platform/frameworks/base/+/badc69af3d741a737900eddd90279ab94e15a88a

https://cs.android.com/android/_/android/platform/frameworks/base/+/f50a293bed5958532697106034b82361ec763d96

https://cs.android.com/android/platform/superproject/+/android-8.0.0_r1:frameworks/base/services/core/java/com/android/server/pm/PackageManagerShellCommand.java;l=700

https://cs.android.com/android/platform/superproject/+/android-10.0.0_r1:frameworks/base/services/core/java/com/android/server/pm/PackageManagerShellCommand.java;l=2997

https://cs.android.com/android/platform/superproject/+/android-13.0.0_r49:frameworks/base/services/core/java/com/android/server/pm/PackageManagerShellCommand.java;l=891

**Comment by 2096779623** [comment]
> p.s The `--show-versioncode` option was added in Android `8` but wasn't shown in help output till android `10`, can be added if needed by checking `ro.build.sdk.version` against `>= 26`.

```bash
escaped_package_name="$(echo -n "@TERMUX_APP_PACKAGE@" | sed -zE -e 's/[][\.|$(){}?+*^]/\\&/g')"
if [ "$(getprop ro.build.version.sdk)" -ge "26" ]; then
    SHOW_VERSION_CODE="--show-versioncode"
fi
TERMUX_PLUGINS="$(pm list packages --user "${TERMUX__USER_ID:-0}" $SHOW_VERSION_CODE 2>&1 </dev/null | grep -E "^package:$escaped_package_name\.[a-zA-Z]" | cut -d ":" -f 2- | grep -vE "^$escaped_package_name\.tapm\.[a-zA-Z]")"
```

**Comment by agnostic-apollo** [comment]
Yeah, that would work, but check output of `getprop` to ensure its an integer before `-ge`, otherwise will trigger the `integer expression expected` error.

```shell
sdk_version="$(getprop ro.build.version.sdk || :)"
if [[ "$sdk_version" =~ ^[0-9]+$ ]] && [ "$sdk_version" -ge "26" ]; then
```

**Comment by agnostic-apollo** [comment]
Updated...

**Comment by agnostic-apollo** [comment]
Looks good, thanks!

**Pull request merged**